File: medium/200.ts
/**
    200. Number of Islands
 */
function numIslands(grid: string[][]): number {
    function find(idx: number): number {
        if (idx != parents[idx])
            idx = find(parents[idx])
        return parents[idx]
    }

    const m: number = grid.length
    const n: number = grid[0].length
    const parents: number[] = new Array(m * n).fill(-1)
    const s: number[] = []
    const dir: number[] = [-1, 0, 1, 0, 0, -1, 0, 1]
    const toIdx = (i: number, j: number) => n * i + j
    let res: number = 0

    for (let i = 0; i < m; ++i)
        for (let j = 0; j < n; ++j)
            if (grid[i][j] == '1') {
                const idx: number = toIdx(i, j)
                parents[idx] = idx
                s.push(i, j)
                ++res
            }
    for (let l = 0; l < s.length; l += 2) {
        const [i, j]: number[] = [s[l], s[l + 1]]
        const p1: number = find(toIdx(i, j))
        for (let k = 0; k < 8; k += 2) {
            const [x, y] = [i + dir[k], j + dir[k + 1]]
            if (x < 0 || y < 0 || x >= m || y >= n || grid[x][y] == '0')
                continue
            const p2: number = find(toIdx(x, y))
            if (p1 != p2) {
                parents[p2] = p1
                --res
            }
        }
    }

    return res
};

File: medium/200.cpp
/**
    # 200. Number of Islands
*/

class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        int m = grid.size(), n = grid[0].size();
        vector<int> root(m * n, -1);

        // init root
        for(int i = 0; i < m; ++i)
            for(int j = 0; j < n; ++j)
                if(grid[i][j] == '1'){
                    int idx = getIdx(i, j, n);
                    root[idx] = idx;
                }

        // union
        for(int i = 0; i < m; ++i)
            for(int j = 0; j < n; ++j){
                int idx = getIdx(i, j, n);
                // union right
                if(j != n - 1)
                    Union(root, idx, idx + 1);
                // union below
                if(i != m - 1)
                    Union(root, idx, idx + n);
            }

        // count island
        unordered_set<int> set;
        for(int i = 0; i < m * n; ++i){
            find(root, i);
            set.insert(root[i]);
        }

        return set.find(-1) != set.end() ? set.size() - 1 : set.size(); // remove -1
    }

private: 
    int getIdx(int x, int y, int n){
        return x * n + y;
    }

    int find(vector<int> &root, int idx){
        if(root[idx] == -1) return -1;
        if(root[idx] != idx)
            root[idx] = find(root, root[idx]);
        return root[idx];
    }

    void Union(vector<int> &root, int x, int y){
        if(root[x] == -1 || root[y] == -1) return;
        int px = find(root, x), py = find(root, y);
        if(px < py) root[py] = root[px];
        else root[px] = root[py];
    }
};
